lmdeploy/turbomind/deploy/module.py

```diff
@@ -0,0 +1,261 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from abc import ABC, abstractmethod
+from functools import partial
+
+import torch
+
+from .parameter import get_params
+from .source_model.base import BaseReader
+from .target_model.base import BaseOutputModel
+
+
+def permute_v2(x: torch.Tensor, size_per_head: int = 128):
+    """
+        Contract: x.size(-1) is output dims
+    """
+
+    assert x.size(-1) > 1
+
+    output_dims = x.size(-1)
+    head_num = output_dims // size_per_head
+
+    return x.view(-1, head_num, 2,
+                  size_per_head // 2).transpose(2, 3).reshape(x.shape)
+
+
+def merge_qkv_v2(q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, tp: int):
+    """
+        Contract: x.size(-1) is output dims
+    """
+
+    def reshape(x):
+        return x.view(x.size(0), tp, -1) if q.dim() == 2 else x.view(tp, -1)
+
+    qkv = torch.cat(tuple(map(reshape, (q, k, v))), dim=-1)
+
+    qkv = qkv.view(-1, qkv.size(-1) * tp)
+    if q.dim() == 1:
+        qkv.squeeze_()
+
+    return qkv
+
+
+def transpose(x):
+    return x.t() if x is not None else x
+
+
+def pad_out_dims(x: torch.Tensor, dims: int):
+    pad = dims - x.size(-1)
+    assert pad >= 0
+    return torch.nn.functional.pad(x, (0, pad), 'constant', 0)
+
+
+def pad_in_dims(x: torch.Tensor, dims: int):
+    pad = dims - x.size(0)
+    assert x.dim() == 2
+    assert pad >= 0
+    return torch.nn.functional.pad(x, (0, 0, 0, pad), 'constant', 0)
+
+
+# split out dims -> copy A, split-out-dims B (qkv, w1, w3)
+# split  in dims -> split-in-dims A,  copy B (  o, w2)
+def get_lora_flags(kind: str):
+    return ('lora_a' in kind, 'lora_b' in kind)
+
+
+class Module(ABC):
+
+    def __init__(self, model: BaseOutputModel):
+        self.model = model
+
+    def __call__(self, *args, **kwargs):
+        return self.apply(*args, **kwargs)
+
+    @abstractmethod
+    def apply(self, r: BaseReader, idx: int):
```

Comment:
May change it to `def apply(self, idx, r)` since all the derived classes put `idx` before `r`